### ISSUE TITLE
Possible issue with slab rebalance.

### ISSUE THREAD

**Issue by danielrushton** [memcached/memcached#151]
I have been looking over the code in slabs.h/slabs.c and trying to understand how the source and destination slabs are chosen for the rebalancing algorithm. 

In slab_rebalance_finish I see that src and destination are set to zero, presumably for the next run.

809  slab_rebal.s_clsid    = 0;
810  slab_rebal.d_clsid    = 0;

These vars are then used to index into the slabs array in  slab_rebalance_alloc, slab_rebalance_move, and slab_rebalance_finish routines, however, I do not see anywhere in which s_clsid or d_clsid are ever set to anything besides 0 (aside from the slab_reassign command). In other words it appears that the source and destination are always set to zero and the conditional on line 508 of slabs.c will always result in no_go = -2.

It does not appear that the values are modified outside of slabs.c either.

`if (slab_rebal.s_clsid < POWER_SMALLEST ||
        slab_rebal.s_clsid > power_largest  ||
        slab_rebal.d_clsid < SLAB_GLOBAL_PAGE_POOL ||
        slab_rebal.d_clsid > power_largest  ||
        slab_rebal.s_clsid == slab_rebal.d_clsid)
        no_go = -2;`

**Comment by dormando** [comment]
Hey,

What version are you looking at? That code got completely redone in 1.4.25

**Comment by danielrushton** [comment]
Hey Dormando, thanks for the quick response. I am looking at v 1.4.25, which I cloned about a week ago. If you do a quick search of slabs.c for s_clsid you'll see what I mean. It is set by the command handler code but not by the slab rebalancing thread. It just remains zero from what I can tell. I grep'd over the entire project looking for where these variables are set.

**Comment by dormando** [comment]
Sorry for the super late response.

You're right in that it's only set in slabs_reassign(). This is a little confusing I guess.

I did a few things in the 1.4.25 release:

 * The LRU juggler watches for too many free chunks in a slab class, which then reassigns a page into the global page pool (items.c):
```
    /* If slab automove is enabled on any level, and we have more than 2 pages
     * worth of chunks free in this class, ask (gently) to reassign a page
     * from this class back into the global pool (0)
     */
    if (settings.slab_automove > 0 && chunks_free > (chunks_perslab * 2.5)) {
        slabs_reassign(slabs_clsid, SLAB_GLOBAL_PAGE_POOL);
    }
```

Once a page has been moved back to SLAB_GLOBAL_PAGE_POOL's magic classid, the slab rebalancer isn't involved at all in reassigning the memory.

When a `set` happens a new chunk of memory is requested, memcached checks for free chunks in the slab class. If none exist, it will attempt to pull a 1MB page from malloc *or the global page pool*, then splitting into chunks.

see: slabs.c:do_slabs_newslab, which calls slabs.c:get_page_from_global_pool. Using the global page pool is very similar to simply free()'ing a page and lowering the memory allocated counter.. just avoids the extra overhead a little.

Feature's in use by a number of people and works. Thanks for checking things over though :) Going to close this out.
